enderio-base/src/main/java/crazypants/enderio/base/handler/darksteel/DarkSteelRecipeManager.java

```diff
@@ -188,6 +187,12 @@ public static void addAdvancedTooltipEntries(@Nonnull ItemStack itemstack, Entit
     return result.isEmpty() ? "" : NullHelper.first(result.substring(1), "");
   }
 
+  public static @Nonnull NNList<IDarkSteelUpgrade> getUpgrades(@Nonnull ItemStack stack) {
+    NNList<IDarkSteelUpgrade> list =  UpgradeRegistry.getUpgrades();
+    list.removeIf(upgrade -> !upgrade.hasUpgrade(stack));
```

Comment:
I worry that this O(n) code is going to slow down the JEI handler significantly. Have you profiled the difference?